contracts/babylon/src/contract.rs

```diff
@@ -309,15 +311,109 @@ pub fn execute(
             // TODO: Add events (#124)
             Ok(res)
         }
+        ExecuteMsg::DistributeRewards { fp_distribution } => {
+            // Check that the sender is the finality contract
+            let cfg = CONFIG.load(deps.storage)?;
+            let finality_addr = cfg
+                .btc_finality
+                .ok_or(ContractError::BtcFinalityNotSet {})?;
+            if info.sender != finality_addr {
+                return Err(ContractError::Unauthorized {});
+            }
+
+            // Check that the funds are there
+            let funds_sent_total = must_pay(&info, &cfg.denom)?;
+
+            // Calculate total rewards to send
+            let total_rewards = fp_distribution
+                .iter()
+                .map(|reward_info| reward_info.reward)
+                .sum();
+
+            // Check that the total rewards  to distribute match the funds sent
+            if funds_sent_total != total_rewards {
+                return Err(ContractError::InvalidRewards(
+                    total_rewards,
+                    funds_sent_total,
+                ));
+            }
+
+            // Get consumer name for bsn_consumer_id
+            let bsn_consumer_id = cfg
+                .consumer_name
+                .ok_or(ContractError::ConsumerNameNotSet {})?;
+
+            // Get the ICS20 transfer channel ID
+            let channel_id = IBC_TRANSFER_CHANNEL.load(deps.storage)?;
+
+            // Calculate ratios from absolute amounts
+            let fp_ratios: Vec<FpRatio> = fp_distribution
+                .iter()
+                .map(|reward_info| {
+                    // Convert to Decimal for proper ratio calculation
+                    let reward_decimal = Decimal::from_ratio(reward_info.reward, total_rewards);
+
+                    let btc_pk_bytes = hex::decode(&reward_info.fp_pubkey_hex)?;
+                    Ok(FpRatio {
+                        btc_pk: Binary::new(btc_pk_bytes),
+                        ratio: reward_decimal,
+                    })
+                })
+                .collect::<Result<Vec<_>, ContractError>>()?;
+
+            // Create memo with the proper Babylon structure using typed structs
+            let callback_memo = CallbackMemo {
+                action: "add_bsn_rewards".to_string(),
```

Comment:
Not an IBC expert, I'm wondering if it's a convention to name the action `add_bsn_rewards` matching the info in `dest_callback` or if it must be named as such; otherwise, the callback won't work.

Reply:
This is part of the (custom) memo field defintion. It has to be set to that value (which could be a constant) for the message to work / be processed.